### ISSUE TITLE
Receiving error undefined prevents Slack client from reconnecting

### ISSUE THREAD

**Issue by stanhu** [slackhq/hubot-slack#206]
It looks like the following sequence happened:

1. Slack client received ENOTFOUND
2. Slack client exited, process restarted by manager
3. Slack client attempted to reconnect but got "Received error defined"
4. In `@robot.logger.error error.stack`, `error.stack` was undefined, leading to an Exception.

Seems to me the timeout to exit the process should be set before the print statements occur.

```
Fri Jun 26 2015 11:05:04 GMT-0700 (PDT)] ERROR Last pong is too old: 14.935
[Fri Jun 26 2015 11:05:04 GMT-0700 (PDT)] INFO Reconnecting in 1000ms
[Fri Jun 26 2015 11:05:05 GMT-0700 (PDT)] INFO Attempting reconnect
[Fri Jun 26 2015 11:05:05 GMT-0700 (PDT)] INFO Connecting...
[Fri Jun 26 2015 11:05:34 GMT-0700 (PDT)] INFO Slack client closed
[Fri Jun 26 2015 11:06:03 GMT-0700 (PDT)] INFO Connecting...
[Fri Jun 26 2015 11:06:59 GMT-0700 (PDT)] ERROR Received error "ENOTFOUND"
[Fri Jun 26 2015 11:06:59 GMT-0700 (PDT)] ERROR undefined
[Fri Jun 26 2015 11:06:59 GMT-0700 (PDT)] ERROR Exiting in 1 second
[Fri Jun 26 2015 11:06:59 GMT-0700 (PDT)] ERROR TypeError: Cannot call method 'close' of null
  at Client.reconnect (/home/pzbot/hubot/node_modules/hubot-slack/node_modules/slack-client/src/client.js:185:13)
  at Client._onLogin (/home/pzbot/hubot/node_modules/hubot-slack/node_modules/slack-client/src/client.js:77:23)
  at /home/pzbot/hubot/node_modules/hubot-slack/node_modules/slack-client/src/client.js:2:59
  at ClientRequest.<anonymous> (/home/pzbot/hubot/node_modules/hubot-slack/node_modules/slack-client/src/client.js:682:18)
  at ClientRequest.emit (events.js:95:17)
  at CleartextStream.socketErrorListener (http.js:1548:9)
  at CleartextStream.emit (events.js:95:17)
  at Socket.onerror (tls.js:1456:17)
  at Socket.emit (events.js:117:20)
  at net.js:834:16
  at process._tickCallback (node.js:448:13)

[Fri Jun 26 2015 11:07:02 GMT-0700 (PDT)] INFO Connecting...
[Fri Jun 26 2015 11:07:50 GMT-0700 (PDT)] INFO Logged in as xxx of YYY, but not yet connected
[Fri Jun 26 2015 11:08:34 GMT-0700 (PDT)] ERROR Received error undefined
[Fri Jun 26 2015 11:08:34 GMT-0700 (PDT)] ERROR TypeError: Cannot read property 'stack' of undefined
  at SlackBot.error (/home/pzbot/hubot/node_modules/hubot-slack/src/slack.coffee:48:5, <js>:66:36)
  at Client.<anonymous> (/home/pzbot/hubot/node_modules/hubot-slack/src/slack.coffee:1:1, <js>:3:61)
  at Client.emit (events.js:92:17)
  at WebSocket.<anonymous> (/home/pzbot/hubot/node_modules/hubot-slack/node_modules/slack-client/src/client.js:145:24)
  at WebSocket.emit (events.js:95:17)
  at ClientRequest.<anonymous> (/home/pzbot/hubot/node_modules/hubot-slack/node_modules/slack-client/node_modules/ws/lib/WebSocket.js:599:10)
  at ClientRequest.emit (events.js:95:17)
  at CleartextStream.socketErrorListener (http.js:1548:9)
  at CleartextStream.emit (events.js:95:17)
  at Socket.onerror (tls.js:1456:17)
  at Socket.emit (events.js:117:20)
  at net.js:441:14
  at process._tickCallback (node.js:448:13)
```

**Comment by technicalpickles** [comment]
> Slack client received ENOTFOUND

Do you have any idea on what caused this? Some quick research points towards not being able to resolve a DNS record (not sure if it's incorrect or just unresponsive).

**Comment by stanhu** [comment]
Yes, DNS on the machine was inaccessible.

**Comment by pixelrebel** [comment]
@technicalpickles I'm having a similar experience, but my hubot is behind a proxy.  

```
[Thu Jun 16 2016 15:32:39 GMT-0700 (PDT)] DEBUG Loading adapter slack
[Thu Jun 16 2016 15:32:39 GMT-0700 (PDT)] INFO Connecting...
[Thu Jun 16 2016 15:32:39 GMT-0700 (PDT)] ERROR undefined
[Thu Jun 16 2016 15:32:39 GMT-0700 (PDT)] INFO Reconnecting in 1000ms
[Thu Jun 16 2016 15:32:40 GMT-0700 (PDT)] INFO Attempting reconnect
[Thu Jun 16 2016 15:32:40 GMT-0700 (PDT)] INFO Connecting...
[Thu Jun 16 2016 15:32:40 GMT-0700 (PDT)] ERROR undefined
```

I've tried both suggestions here without luck:
https://hubot.github.com/docs/patterns/#forwarding-all-http-requests-through-a-proxy
http://tomute.hateblo.jp/entry/2013/07/13/233258

DNS resolution is working for because my squid proxy server doubles as a DNS relay.  Any tips for me?

**Comment by DEGoodmanWilson** [comment]
Going to start looking into this soon, just a heads-up!

**Comment by baudehlo** [comment]
Any progress on this? Our hubot slack bot just hangs at startup.

```
2016-07-12T14:53:58.97602 [Tue Jul 12 2016 14:53:58 GMT+0000 (UTC)] INFO Connecting...
2016-07-12T14:53:59.08744 [Tue Jul 12 2016 14:53:59 GMT+0000 (UTC)] ERROR undefined
```

And then nothing.

**Comment by baudehlo** [comment]
Did some debugging - it was getting a 429 error on connect (rate limited). I think it must be a bug in slackhq because that bot had been disconnected due to rate limits for days.

I regenerated the auth token and it's working again.

**Comment by DEGoodmanWilson** [comment]
This is an issue that should be resolved in the next release, which has much improved reconnection logic—you can preview that here: https://github.com/slackhq/hubot-slack/pull/309

**Comment by DEGoodmanWilson** [comment]
I'm going to close this as fixed by v4.0, unless and until someone reports otherwise!

**Comment by alex-moreno** [comment]
we had this reconnecting issue, simply regenerating the API token fixes it
